### PR TITLE
Sipify QgsFields

### PR DESCRIPTION
**Pull request by arnaud-morvan**
## Description
Include a few sentences describing the overall goals for this PR (pull request). If applicable also add screenshots.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: python/core/qgsfields.sip [hunk 1]
```diff
@@ -1,253 +1,332 @@
...
-    void extend( const QgsFields& other );
+    void extend( const QgsFields &other );
+%Docstring
+Extend with fields from another QgsFields container
+%End
 
-    //! Check whether the container is empty
     bool isEmpty() const;
-    //! Return number of items
-    int count() const;
```
Comment (nyalldawson) [inline python/core/qgsfields.sip]: Count seems to have dropped off here.

Comment (3nids) [inline python/core/qgsfields.sip]: there is the `__len__` method instead, so I believe it's fine?

Comment (nyalldawson) [inline python/core/qgsfields.sip]: Sure, but in that case it'll need adding to the API break docs, and checking over the python plugin source to make sure it's not used there.

Comment (3nids) [inline python/core/qgsfields.sip]: I don't mind keeping then. Any opinion, ok to put it back @arnaud-morvan ?

Comment (nyalldawson) [inline python/core/qgsfields.sip]: There's also size - if we keep/remove count, we should do the same with size. Personally I like adding the len option to python, but would prefer to keep the underlying c++ methods available still.

Comment (arnaud-morvan) [inline python/core/qgsfields.sip]: I've seen this in old sip:
```
//int count() const /__len__/;
```
but the count method was really there just before, I'll put it back.

File: python/core/qgsfields.sip [hunk 2]
```diff
@@ -1,253 +1,333 @@
...
+      sipRes = new QIcon( sipCpp->iconForField( a0 ) );
+    }
+%End
+
     operator QVariant() const;
+%Docstring
+Allows direct construction of QVariants from fields.
+%End
 
-/*  SIP_PYOBJECT __getitem__(int key);
```
Comment (3nids) [inline python/core/qgsfields.sip]: it seems __getime__ is missing in the sipified file

Comment (arnaud-morvan) [inline python/core/qgsfields.sip]: This was in comment and was falling back in a dockstring. This should be handled by operator[].

Comment (arnaud-morvan) [inline python/core/qgsfields.sip]: I will add a test for the operator[] on python side.

**Pull request merged**